### DISCUSSION TITLE
How do I use an uploaded file?

### DISCUSSION THREAD

**Discussion by ntatko** [directus/directus#4622] [Archive]
Howdy!

I'm looking to use files uploaded to directus in a web site. You know, `<img src="some url" />`, but I can't for the life of me figure out how to allow this.

The docs are a little sparse on this (https://docs.directus.io/guides/files/#downloading-an-original-file).

Where or how can I find the absolute URL for a media file? 

(p.s. uploading images via API is also rather undocumented... but I'll cross that bridge later)

**Comment by rijkvanzanten** [comment]
See https://docs.directus.io/reference/api/assets/ 🙂 

As for file uploads, have you seen https://docs.directus.io/reference/api/rest/files/#upload-a-file ?

**Reply by drodenberg** [reply]
How do you make that endpoint publicly accessible? I dont see it in the public roles and permissions section

**Reply by rijkvanzanten** [reply]
The file-read permissions are based on the permissions for directus_files currently 👍🏻

**Reply by drodenberg** [reply]
<img width="650" alt="Screen Shot 2021-03-22 at 5 39 06 PM" src="https://user-images.githubusercontent.com/5175281/112067159-9ab4cd80-8b35-11eb-9ca7-8d69f2e091f6.png">
This is what I see under our permission? Sorry, I'm not following how to set the `read-file` permissions. I have the `Directus Files` set to public everything but when I hit an assets endpoint with a file id that we have stored i get a `You don't have permission to access this.` error

**Reply by drodenberg** [reply]
Ok, I can get one of them to work like 30% of the time. When I refresh it will say no permissions but every third time i refresh it will show the image

**Reply by rijkvanzanten** [reply]
That sounds odd.. Where are you retrieving that asset? Do you have any caching/cdn magic in between?

**Reply by drodenberg** [reply]
Its stored on our directus app. We aren't doing anything out of the ordinary that I know of.

**Reply by rijkvanzanten** [reply]
I meant, where are you seeing this behavior:

> When I refresh it will say no permissions but every third time i refresh it will show the image

**Reply by drodenberg** [reply]
Oh, when I'm hitting the /assets/id route

**Reply by drodenberg** [reply]
https://user-images.githubusercontent.com/5175281/112068882-cc7b6380-8b38-11eb-9071-70a651cb3922.mov

**Reply by rijkvanzanten** [reply]
This on localhost w/ local file storage?

**Reply by drodenberg** [reply]
Its on our hosted app of directus

**Reply by drodenberg** [reply]
I do notice that in the file viewer in directus the storage area input does say `local`... is that maybe the problem?

**Reply by rijkvanzanten** [reply]
Where / how is it hosted? Where is it pulling the files from? 

Could it be that you have a load balanced instance with multiple copies of the process, where the processes aren't pulling from the same file storage? That would explain the odd behavior (if the file only exists in 1/3 load balanced instances, you'd expect a ~33% success rate 🤔 )

**Reply by drodenberg** [reply]
That's a good thought for sure. I'll check with my team. I'm not sure how the backend is all setup. Thanks a ton for the help.

**Reply by rijkvanzanten** [reply]
No problem! For multi-process hosted instances, I'd definitely configuring Directus to save to / pull from an external object storage service like S3, Google Cloud Storage, or Azure Blob Storage. That'll get rid of this problem as well 🙂  (see https://docs.directus.io/reference/environment-variables/#file-storage for all configuration flags for that)

**Reply by ntatko** [reply]
We're using local storage - it's running in docker on a linode instance (linux VM) with 60gb of storage. I've considered using straight up S3 as I've got working code, but if we could tie into directus directly (lol), it would save us some trouble.

**Reply by rijkvanzanten** [reply]
I'm assuming that Linode is running that docker image in parallel then, causing this behavior. 

What do you mean by 

> but if we could tie into directus directly

? 

Directus has support for s3/gcs/azure for file storage out of the box, if that's what you're referring to :)
